Draft - 254 check translations

mrsarm opened the pull request:
Ticket: https://github.com/medic/medic-conf/issues/254

Add to the POE script the same validations from the new module [translation-checker](https://github.com/medic/translation-checker/pull/1) that the one added in this [PR](https://github.com/medic/medic-conf/issues/254) to the medic-conf tool. 

- Fails if "messageformat" errors or bad usage of placeholders are found
- Shows how many empty translations were found but without failing
- The old script under the `scripts/bad_translations/` folder was removed

This is the how the errors are displayed now:

```
$ ./cli.js export .
  ◡ Exporting translations...{"response":{"status":"success","code":"200","message":"OK"},"result":{"languages":[{"name":"Bambara","code":"b...}
  ◠ Exporting translations...   hi saved to /projects/medic/cht-core/scripts/poe/./messages-hi.properties
  ◟ Exporting translations...   en saved to /projects/medic/cht-core/scripts/poe/./messages-en.properties...
...
...
...
  ◡ Exporting translations...Error: Cannot compile 'bm' translation Confirm delete = 'I bafɛ ka nin tɔgɔ {{}} dɔsi wa? I tese ka a lasegi tuguni' : Expected identifier but "{" found.
Error: Cannot compile 'es' translation n.month = '{MONTHS, plural, uno{1 mes} otros{# meses}}
' : Invalid key `uno` for argument `MONTHS`. Valid plural keys for this locale are `one`, `other`, and explicit keys like `=0`.
Error: Cannot compile 'es' translation n.week = '{WEEKS, plural, uno{1 semana} otros{# semanas}}' : Invalid key `uno` for argument `WEEKS`. Valid plural keys for this locale are `one`, `other`, and explicit keys like `=0`.
Error: Cannot compile 'id' translation contacts.visits.visits = '{VISITS, plural, one{visit} other{visits}}' : Invalid key `one` for argument `VISITS`. Valid plural keys for this locale are `other`, and explicit keys like `=0`.
Error: Cannot compile 'sw' translation field digits only = '{field}} lazima tu tarakimu namba' : Expected "#", "{", doubled apostrophe, end of input, escaped string, or plain char but "}" found.
Error: Cannot compile 'sw' translation messages.c.validation.weeks_since_birth = 'format ya usajili kwa ajili {{}} patient_id si sahihi, tafadhali kuhakikisha kuwa umri si katika wiki na kati ya 0 na 90.' : Expected identifier but "{" found.
Info: There were 5794 empty translations trying to compile
Error: There were 6 errors trying to compile
Error: Invalid placeholders or "messageformat" messages!
    at download (/projects/medic/cht-core/scripts/poe/lib/poe.js:66:13)
    at <anonymous>
```

garethbowen commented on config/standard/translations/messages-en.properties:
```diff
@@ -1,3 +1,6 @@
+messages.d.validation.delivery_code = {{\#delivery_code}}The delivery code was not understood. Please use code 'F' for facility, 'S' for homebirth with Skilled Birth Attendant, 'NS' for birth without SBA.{{/delivery_code}}{{^delivery_code}}The delivery code was missing. Please use code 'F' for facility, 'S' for homebirth with Skilled Birth Attendant, 'NS' for birth with no skilled attendant.{{/de ...
+messages.c.validation.patient_name = {{\#patient_name}}Registration format is incorrect. Start with C followed by space, age in weeks followed by a space, and the child's name (maximum of 30 characters).{{/patient_name}}{{^patient_name}}.The registration format is missing the name. Start with C followed by space, age in weeks followed by a space, and the child's name{{/patient_name}}.
+messages.n.validation.patient_name = {{\#patient_name}}The registration format is incorrect, ensure the message starts with N followed by space and the person's name (maximum of 30 characters).{{/patient_name}}{{^patient_name}}The registration format is incorrect, ensure the message starts with N followed by space and the person's name.{{/patient_name}}.
```
Where did these come from? :thinking:

mrsarm replied:
https://github.com/medic/cht-core/blob/master/ddocs/medic/_attachments/translations/messages-en.properties#L760 :slightly_smiling_face:

garethbowen replied:
Ok. They shouldn't be listed in both files. The ones in `_attachments` are automatically deployed for each instance anyway. From the look of these keys they should _only_ be specified in the `config/standard/translations/` file, so I guess remove them from the `ddocs/medic/_attachments/translations` one?

mrsarm replied:
Yes I added the 'en' translations because these keys are used in [messages-sw.properties](https://github.com/medic/cht-core/blob/125e1c8b3c9c9f4155a0e3572d8f734b9a0c0f61/config/standard/translations/messages-sw.properties#L1-L3) and have placeholders that need to much the base translations, so as you suggest I will remove them in both 'en' and 'sw' files, I checked the 'sw' texts and there is no difference with the ones in `_attachments/translations/messages-sw.properties`.

mrsarm replied:
Done.

garethbowen commented on scripts/poe/lib/validate.js:
```diff
@@ -1,21 +1,31 @@
+const transErrorsMsg = MFORMAT
+  .compile('There {ERRORS, plural, one{was 1 error} other{were # errors}} trying to compile');
+const transEmptyMsg = MFORMAT
+  .compile('There {EMPTIES, plural, one{was 1 empty translation} other{were # empty translations}} trying to compile');
```
Now that this is the only use of `MessageFormat` it'd be better to drop it altogether and just using hardcoded strings for error reporting.

garethbowen commented on scripts/poe/package.json:
```diff
@@ -20,13 +20,15 @@
-    "request": "^2.88.0"
+    "messageformat": "^2.3.0",
+    "request": "^2.88.0",
+    "translation-checker": "file:../../../translation-checker"
```
Don't forget to change this once the translation-checker is published!

mrsarm replied:
:man_facepalming: Oh yes in fact the intention was to commit "1.0.0" like in the medic-conf [PR](https://github.com/medic/medic-conf/blob/254-medic-conf-translation-check/package.json#L60), I'll change it.

mrsarm commented:
@garethbowen ,

> I think the scripts/poe/package-lock.json needs to be cleaned up - it has thousands of changes I don't think are required.

I have updated all the dependencies because they were really outdated and with a lot of security vulnerabilities reported, here is the output when you install the master version:

```
$ npm install
npm WARN poe@1.0.0 No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 661 packages from 399 contributors and audited 729 packages in 7.124s
found 1078 vulnerabilities (738 low, 15 moderate, 323 high, 2 critical)
  run `npm audit fix` to fix them, or `npm audit` for details
```

Do you want me to rollback the changes?

garethbowen commented:
> Do you want me to rollback the changes?

No that's fine - I just wanted to make sure those were intentional changes. Nice one!

garethbowen commented:
Approved. Looks like there's a flaky test which has failed a few of your builds. Issue raised here: https://github.com/medic/cht-core/issues/6605
